Add initial IPv6 documentation (#12649)
kubernetes/kops: docs/networking/ipv6.md

# IPv6

{{ kops_feature_table(kops_added_ff='1.22') }}

kOps has experimental support for configuring clusters with IPv6-only pods and dual-stack nodes.

IPv6 mode is specified by setting `nonMasqueradeCIDR: "::/0"` in the cluster spec.
The `--ipv6` flag of `kops create cluster` sets this field, among others.

## Cloud providers

kOps currently supports IPv6 on AWS only.

# VPC and subnets

The VPC can be either shared or managed by kOps. If shared, it must have an IPv6 pool associated.

Subnet IPv6 CIDR allocations may be specified in the cluster spec using the special syntax "/LEN#N",
where "LEN" is the prefix length and "N" is the hexadecimal sequence number of the CIDR within the VPC's IPv6 CIDR.
For example, if the VPC's CIDR is "2001:db8::/56" then the syntax "/64#a" would mean "2001:db8:0:a/64".

# CNI

kOps currently supports IPv6 on Calico, Cilium, and bring-your-own CNI only.

CNIs must not masquerade IPv6 addresses.

## Calico

Running IPv6 with Calico requires a Debian 11-based AMI. As of the writing of this document, Ubuntu does not work due to an 

> [issue with systemd's handling of AWS's incorrect DHCP responses](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1945888).

Author: Do we have a public ticket reference for the AWS DHCP bug?  Do we want to describe it in the "Cloud providers" section?

Reviewer: Not that I know of.

> [issue with systemd's handling of AWS's incorrect DHCP responses](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1945888).

Reviewer: https://github.com/systemd/systemd/issues/20803 may be a better reference.

Author: I was trying to go for something Ubuntu-specific, but I suppose AL2 would probably have the same issue.

Reviewer: I don't know if AL2 uses systemd-networkd. Each distro uses something different. I Just mention Debian 11 if it works with Cilium, and mention that Ubuntu 22.04 should work too.

Author: Perhaps you could refine the language after the PR lands? I don't have complete enough knowledge of all the possibilities.

Reviewer: 👍

# IPv6

> {{ kops_feature_table(kops_added_ff='1.22') }}

Reviewer: Best not to confuse people:
```suggestion
{{ kops_feature_table(kops_added_ff='1.23') }}
```

Author: We did feature flag it in 1.22

Reviewer: Yes, but it's a totally different approach that we don't want to document or support. Mentioning 1.22 would confuse some users in thinking that it's same as what we will release in 1.23. Am I wrong?

Author: I suppose it depends on what the table is supposed to mean and how people interpret "experimental". I suspect there's a good chance we'll remove the feature flag for 1.23.

Reviewer: I agree that the feature flag will probably be removed in 1.23. Though, kOps docs don't have a way to say, experimental in version X, beta in version Y, stable in version Z. I am afraid people will try this now with 1.22 and get something different than described here.

> kOps has experimental support for configuring clusters with IPv6-only pods and dual-stack nodes.

Reviewer: Dual-stack should probably not get mentioned.
```suggestion
kOps has experimental support for configuring clusters with IPv6-only pods on AWS instances with IPv6 enabled.
```

Author: The nodes are dual-stack.
